plane/src/typed_unix_socket/client.rs

```diff
@@ -0,0 +1,247 @@
+use super::{get_quick_backoff, WrappedMessage};
+use crate::util::random_token;
+use anyhow::Error;
+use dashmap::DashMap;
+use serde::{Deserialize, Serialize};
+use std::{clone::Clone, fmt::Debug, path::Path, sync::Arc};
+use tokio::io::AsyncWriteExt;
+use tokio::io::{AsyncBufReadExt, BufReader, BufWriter};
+use tokio::{
+    net::UnixStream,
+    sync::{broadcast, oneshot, watch},
+};
+
+#[derive(Clone)]
+pub struct TypedUnixSocketClient<MessageToServer, MessageToClient>
+where
+    MessageToServer: Send + Sync + 'static + Clone + Debug + Serialize + for<'de> Deserialize<'de>,
+    MessageToClient: Send + Sync + 'static + Clone + Debug + Serialize + for<'de> Deserialize<'de>,
+{
+    tx: broadcast::Sender<WrappedMessage<MessageToServer>>,
+    response_map: Arc<DashMap<String, oneshot::Sender<MessageToClient>>>,
+    event_tx: broadcast::Sender<MessageToClient>,
+    shutdown_tx: watch::Sender<()>,
+}
+
+impl<MessageToServer, MessageToClient> TypedUnixSocketClient<MessageToServer, MessageToClient>
+where
+    MessageToServer: Send + Sync + 'static + Clone + Debug + Serialize + for<'de> Deserialize<'de>,
+    MessageToClient: Send + Sync + 'static + Clone + Debug + Serialize + for<'de> Deserialize<'de>,
+{
+    pub async fn new<P: AsRef<Path>>(socket_path: P) -> Result<Self, Error> {
+        let (tx, _) = broadcast::channel(100);
+        let response_map = Arc::new(DashMap::new());
+        let (event_tx, _) = broadcast::channel(100);
+        let (shutdown_tx, shutdown_rx) = watch::channel(());
+
+        tokio::spawn({
+            let socket_path = socket_path.as_ref().to_path_buf();
+            let tx = tx.clone();
+            let response_map = Arc::clone(&response_map);
+            let event_tx = event_tx.clone();
+            let mut shutdown_rx = shutdown_rx.clone();
+            async move {
+                let mut backoff = get_quick_backoff();
+                loop {
+                    match UnixStream::connect(&socket_path).await {
+                        Ok(stream) => {
+                            handle_connection(
+                                stream,
+                                tx.subscribe(),
+                                Arc::clone(&response_map),
+                                event_tx.clone(),
+                                shutdown_rx.clone(),
+                            )
+                            .await
+                            .unwrap_or_else(|e| {
+                                tracing::error!("Error handling connection: {}", e);
+                            });
+                            break;
+                        }
+                        Err(e) => {
+                            tracing::error!("Error connecting to server: {}", e);
+                            backoff.wait().await;
```

Comment:
What happens if it never connects? This loop will never break, and these values will never be dropped. Maybe the `backoff` could wait for a maximum amount of time, after which it returns an `Err`?